### PR TITLE
filter chain match: support source CIDRs and ports

### PR DESCRIPTION
**Pull request by mattklein123**
This PR also fully deprecates the tcp_proxy v1 configuration.
This will be deleted following the standard deprecation cycle.
All new uses should use filter chain matching.

Fixes https://github.com/envoyproxy/envoy/issues/4457

Risk Level: Medium
Testing: New tests
Docs Changes: Added
Release Notes: Added
Deprecation Notes: Added

### CODE REVIEW

**Comment by mattklein123** [comment]
@PiotrSikora @htuch PTAL

@htuch This PR contains a fix for the docs build on master. Feel free to pull it out into a different PR in the morning if you want since this might take a few iterations.

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
+
+void ListenerImpl::addFilterChainForSourcePorts(SourcePortsMapSharedPtr& source_ports_map_ptr,
+                                                uint32_t source_port,
+                                                const Network::FilterChainSharedPtr& filter_chain) {
+  if (source_ports_map_ptr == nullptr) {
+    source_ports_map_ptr = std::make_shared<SourcePortsMap>();
+  }
+  auto& source_ports_map = *source_ports_map_ptr;
+
+  if (source_ports_map.find(source_port) != source_ports_map.end()) {
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: nit: avoid double lookup by
```if (!source_ports_map.insert(source_port, filter_chain).second) {```
and 
deleting
```source_ports_map[source_port] = filter_chain;``

File: source/server/listener_manager_impl.cc [hunk 2]
```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
+namespace {
+
+// Template function for creating a CIDR list entry for either source or destination address.
+template <class T>
+std::pair<T, std::vector<Network::Address::CidrRange>> makeCidrListEntry(const std::string& cidr,
+                                                                         const T& data) {
+  std::vector<Network::Address::CidrRange> subnets;
+  if (cidr == EMPTY_STRING) {
+    if (Network::Address::ipFamilySupported(AF_INET)) {
+      subnets.push_back(Network::Address::CidrRange::create("0.0.0.0/0"));
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: IMHO `CidrRange("0.0.0.0/0")` deserves a const as `EMPTY_STRING`

File: source/server/listener_manager_impl.cc [hunk 3]
```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
+    }
+    if (Network::Address::ipFamilySupported(AF_INET6)) {
+      subnets.push_back(Network::Address::CidrRange::create("::/0"));
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: and here

File: source/server/listener_manager_impl.cc [hunk 4]
```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
-void ListenerImpl::convertDestinationIPsMapToTrie() {
+}; // namespace
+
+void ListenerImpl::convertIPsToTries() {
   for (auto& port : destination_ports_map_) {
+    // These variables are used as we build up the destination CIDRs used for the trie.
     auto& destination_ips_pair = port.second;
     auto& destination_ips_map = destination_ips_pair.first;
-    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>> list;
+    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>>
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: save a few copies if we add `destination_ips_list.reserve(destination_ips_map.size())`

File: source/server/listener_manager_impl.cc [hunk 5]
```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
+
+      // This hugely nested for loop greatly pains me, but I'm not sure how to make it better.
+      // We need to get access to all of the source IP strings so that we can convert them into
+      // a trie like we did for the destination IPs above.
+      for (auto& server_names_entry : *entry.second) {
+        for (auto& transport_protocols_entry : server_names_entry.second) {
+          for (auto& application_protocols_entry : transport_protocols_entry.second) {
+            for (auto& source_array_entry : application_protocols_entry.second) {
+              auto& source_ips_map = source_array_entry.first;
+              std::vector<
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: and here: reserve()

File: source/server/listener_manager_impl.cc [hunk 6]
```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
+  return std::make_pair<T, std::vector<Network::Address::CidrRange>>(T(data), std::move(subnets));
 }
 
-void ListenerImpl::convertDestinationIPsMapToTrie() {
+}; // namespace
+
+void ListenerImpl::convertIPsToTries() {
   for (auto& port : destination_ports_map_) {
+    // These variables are used as we build up the destination CIDRs used for the trie.
     auto& destination_ips_pair = port.second;
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: What is the plan for c++17?

File: source/server/listener_manager_impl.cc [hunk 7]
```diff
@@ -580,34 +665,77 @@ const Network::FilterChain*
...
+
+    // Try port 0 if we didn't already try it (UDS).
+    if (source_port != 0) {
+      const auto any_match = source_ports_map.find(0);
+      if (any_match != source_ports_map.end()) {
+        return any_match->second.get();
+      }
+    }
+
+    return nullptr;
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: nit: it's not necessary

File: source/server/listener_manager_impl.cc [hunk 8]
```diff
@@ -580,34 +665,77 @@ const Network::FilterChain*
...
+  static const auto& fake_address = Network::Utility::parseInternetAddress("255.255.255.255");
+
+  auto address = socket.remoteAddress();
+  if (address->type() != Network::Address::Type::Ip) {
+    address = fake_address;
+  }
+
+  // Match on both: exact IP and wider CIDR ranges using LcTrie.
+  const auto& data = source_ips_trie.getData(address);
+  if (!data.empty()) {
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: nit: prefer return null when it's empty to avoid below indent

File: source/server/listener_manager_impl.cc [hunk 9]
```diff
@@ -272,6 +266,12 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
       }
     }
 
+    std::vector<std::string> source_ips;
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: source_ips.reserve(filter_chain_match.source_prefix_ranges())

File: source/server/listener_manager_impl.cc (hunk 6 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: Not sure, but let's discuss in a different context.

File: source/server/listener_manager_impl.cc [hunk 10]
```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
+    if (Network::Address::ipFamilySupported(AF_INET)) {
+      subnets.push_back(Network::Address::CidrRange::create("0.0.0.0/0"));
+    }
+    if (Network::Address::ipFamilySupported(AF_INET6)) {
+      subnets.push_back(Network::Address::CidrRange::create("::/0"));
+    }
+  } else {
+    subnets.push_back(Network::Address::CidrRange::create(cidr));
+  }
+  return std::make_pair<T, std::vector<Network::Address::CidrRange>>(T(data), std::move(subnets));
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: Yes

File: source/server/listener_manager_impl.cc [hunk 11]
```diff
@@ -580,34 +665,77 @@ const Network::FilterChain*
...
                        FilterChainMatch_ConnectionSourceType_EXTERNAL];
 
-  // isLocalConnection can be expensive. Call it only if LOCAL or EXTERNAL are defined.
-  const bool is_local_connection = (filter_chain_local || filter_chain_external)
-                                       ? Network::Utility::isLocalConnection(socket)
-                                       : false;
+  // isLocalConnection can be expensive. Call it only if LOCAL or EXTERNAL have entries.
+  const bool is_local_connection =
+      (!filter_chain_local.first.empty() || !filter_chain_external.first.empty())
+          ? Network::Utility::isLocalConnection(socket)
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: I don't think this avoids the short circuit, either way this code was pre-existing and I don't want to reason about whether the new version is correct or not.

File: api/envoy/api/v2/listener/listener.proto [hunk 12]
```diff
@@ -64,6 +64,8 @@ message Filter {
 // 4. Transport protocol.
 // 5. Application protocols (e.g. ALPN for TLS protocol).
 // 6. Source type (e.g. any, local or external network).
+// 7. Source IP address.
```
Comment (htuch) [inline api/envoy/api/v2/listener/listener.proto]: Can you comment on the rationale behind the ordering here? Is it arbitrary? Is there a practical reason to select source IP/port as the LSBs?

File: api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto [hunk 13]
```diff
@@ -134,46 +119,8 @@ message TcpProxy {
...
-  //
-  // .. attention::
-  //
-  //   In the deprecated V1 configuration, source and destination CIDR ranges are specified as a
-  //   list of strings with each string in CIDR notation. Source and destination ports are
-  //   specified as single strings containing a comma-separated list of ports and/or port ranges.
-  //
-  // Deprecation pending https://github.com/envoyproxy/envoy/issues/4457
-  DeprecatedV1 deprecated_v1 = 6;
+  // [#not-implemented-hide:] Deprecated.
```
Comment (htuch) [inline api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto]: Should `protodoc` have a `[#deprecated:]` annotation that does this? Probably easy to add.

File: source/server/listener_manager_impl.cc [hunk 14]
```diff
@@ -340,67 +342,82 @@ void ListenerImpl::addFilterChain(
...
 
 void ListenerImpl::addFilterChainForServerNames(
-    ServerNamesMap& server_names_map, const std::vector<std::string>& server_names,
+    ServerNamesMapSharedPtr& server_names_map_ptr, const std::vector<std::string>& server_names,
     const std::string& transport_protocol, const std::vector<std::string>& application_protocols,
     const envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType source_type,
+    const std::vector<std::string>& source_ips,
+    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
+  if (server_names_map_ptr == nullptr) {
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: This seems an unrelated change?

File: source/server/listener_manager_impl.cc [hunk 15]
```diff
@@ -410,56 +427,129 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
+namespace {
+
+// Template function for creating a CIDR list entry for either source or destination address.
+template <class T>
+std::pair<T, std::vector<Network::Address::CidrRange>> makeCidrListEntry(const std::string& cidr,
+                                                                         const T& data) {
+  std::vector<Network::Address::CidrRange> subnets;
+  if (cidr == EMPTY_STRING) {
+    if (Network::Address::ipFamilySupported(AF_INET)) {
+      static const std::string catch_all_ipv4 = "0.0.0.0/0";
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: Static initialization order fiasco?

File: source/server/listener_manager_impl.cc [hunk 16]
```diff
@@ -410,56 +427,129 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
...
-        if (Network::Address::ipFamilySupported(AF_INET)) {
-          subnets.push_back(Network::Address::CidrRange::create("0.0.0.0/0"));
-        }
-        if (Network::Address::ipFamilySupported(AF_INET6)) {
-          subnets.push_back(Network::Address::CidrRange::create("::/0"));
+      destination_ips_list.push_back(makeCidrListEntry(entry.first, entry.second));
+
+      // This hugely nested for loop greatly pains me, but I'm not sure how to make it better.
+      // We need to get access to all of the source IP strings so that we can convert them into
+      // a trie like we did for the destination IPs above.
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: Can't we do this more locally, once we've recursed down to source address levels?

File: test/server/listener_manager_impl_test.cc [hunk 17]
```diff
@@ -1596,6 +1546,95 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, SingleFilterChainWithSourceTypeMa
...
+      dynamic_cast<Extensions::TransportSockets::Tls::SslSocket*>(transport_socket.get());
+  auto server_names = ssl_socket->dnsSansLocalCertificate();
+  EXPECT_EQ(server_names.size(), 1);
+  EXPECT_EQ(server_names.front(), "server1.example.com");
+
+  // UDS client. No match.
+  filter_chain =
+      findFilterChain(0, "/tmp/test.sock", "", "tls", {"h2", "http/1.1"}, "/tmp/test.sock", 0);
+  ASSERT_EQ(filter_chain, nullptr);
+}
```
Comment (htuch) [inline test/server/listener_manager_impl_test.cc]: Should we have IPv6 as well? No need if we already are testing this aspect elsewhere..

File: api/envoy/api/v2/listener/listener.proto (hunk 12 above)
Comment (mattklein123) [reply api/envoy/api/v2/listener/listener.proto]: I talked to @PiotrSikora about this offline. Effectively, it's arbitrary, and since we are adding new ones it seems best to add them at the end? IMO this needs to be configurable but that's out of scope for this change. @PiotrSikora any additional thoughts here?

File: api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto (hunk 13 above)
Comment (mattklein123) [reply api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto]: Yes that sounds good. I can work on this in a follow up if that is OK.

File: source/server/listener_manager_impl.cc (hunk 14 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: It's not actually unrelated. To make a long story short, the old code was:
1) Doing a copy
2) More importantly by doing the copy, it made using unique_ptr impossible for nested structures. This was not a problem previously but is a problem now with deeper nesting.

I could probably split this out into a different change, but I would rather not as I don't think it would be a trivial split.

File: source/server/listener_manager_impl.cc (hunk 15 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: I think this is only initialized the first time this code is run so I don't think so?

File: source/server/listener_manager_impl.cc (hunk 16 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: How do you mean? Like break it out into a different function?

File: test/server/listener_manager_impl_test.cc (hunk 17 above)
Comment (mattklein123) [reply test/server/listener_manager_impl_test.cc]: Sure will add an IPv6 test.

File: source/server/listener_manager_impl.cc (hunk 15 above)
Comment (htuch) [reply source/server/listener_manager_impl.cc]: Yeah, that works around the initialization order fiasco, but there is a closely related destruction order problem (https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2). Maybe add CIDR equivalents of `Utility::getIpv4AnyAddress()` etc.

File: source/server/listener_manager_impl.cc (hunk 14 above)
Comment (htuch) [reply source/server/listener_manager_impl.cc]: Fair.

File: api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto (hunk 13 above)
Comment (htuch) [reply api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto]: Ack

File: source/server/listener_manager_impl.cc (hunk 16 above)
Comment (htuch) [reply source/server/listener_manager_impl.cc]: I was thinking that we could build up the trie as we built up the non-optimized version, but it's probably even more complicated than this two phase approach. Fine to keep as is.

File: source/server/listener_manager_impl.cc [hunk 18]
```diff
@@ -580,34 +670,77 @@ const Network::FilterChain*
...
+  } else {
+    return nullptr;
+  }
+}
+
+const Network::FilterChain*
+ListenerImpl::findFilterChainForSourceIpAndPort(const SourceIPsTrie& source_ips_trie,
+                                                const Network::ConnectionSocket& socket) const {
+  // Use invalid IP address (matching only filter chains without IP requirements) for UDS.
+  static const auto& fake_address = Network::Utility::parseInternetAddress("255.255.255.255");
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: This is harder to read than it probably needs to be, with the mix of `static`, ref and const shared ptr.

File: source/server/listener_manager_impl.cc (hunk 15 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: Strict adherence to this rule continues to feel like busy work, but will fix.

File: source/server/listener_manager_impl.cc (hunk 18 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: This was copied from the existing code, will fix.

File: source/server/listener_manager_impl.cc [hunk 19]
```diff
@@ -484,15 +574,23 @@ ListenerImpl::findFilterChain(const Network::ConnectionSocket& socket) const {
   return nullptr;
 }
 
+namespace {
+
+// Return a fake address for use when either the source or destination is UDS.
+Network::Address::InstanceConstSharedPtr fakeAddress() {
+  static const Network::Address::InstanceConstSharedPtr address =
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: Technically shared pointers are non-POD..

File: source/server/listener_manager_impl.cc [hunk 20]
```diff
@@ -580,34 +678,74 @@ const Network::FilterChain*
...
+    return port_match->second.get();
+  }
+
+  // Try port 0 if we didn't already try it (UDS).
+  if (source_port != 0) {
+    const auto any_match = source_ports_map.find(0);
+    if (any_match != source_ports_map.end()) {
+      return any_match->second.get();
+    }
+  }
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: Did you consider making this more regular with the destination IP/port matches to support later reordering?

File: source/server/listener_manager_impl.cc (hunk 19 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: There are many cases of this already in the code base, and there is no way this is an actual issue, but sure, will fix.

File: source/server/listener_manager_impl.cc (hunk 20 above)
Comment (mattklein123) [reply source/server/listener_manager_impl.cc]: What do you mean?

File: test/server/listener_manager_impl_test.cc [hunk 21]
```diff
@@ -1633,6 +1583,137 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, SingleFilterChainWithSourceTypeMa
...
+  EXPECT_CALL(listener_factory_, createListenSocket(_, _, _, true));
+  manager_->addOrUpdateListener(parseListenerFromV2Yaml(yaml), "", true);
+  EXPECT_EQ(1U, manager_->listeners().size());
+
+  // IPv6 client. match.
+  auto filter_chain = findFilterChain(1234, "127.0.0.1", "", "tls", {},
+                                      "2001:0db8:85a3:0000:0000:8a2e:0370:7334", 111);
+  EXPECT_NE(filter_chain, nullptr);
+
+  // IPv6 client. No match.
```
Comment (PiotrSikora) [inline test/server/listener_manager_impl_test.cc]: Nit: This comment is the same as the one above, could you note the difference? e.g. "IPv6 client with matching subnet" / "IPv6 client with non-matching subnet".

File: test/server/listener_manager_impl_test.cc [hunk 22]
```diff
@@ -2238,29 +2303,6 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, MultipleFilterChainsWithSameMatch
                             "the same matching rules are defined");
 }
 
-TEST_F(ListenerManagerImplWithRealFiltersTest,
```
Comment (PiotrSikora) [inline test/server/listener_manager_impl_test.cc]: This test should stay, just change `source_prefix_ranges` to `address_suffix` or `suffix_len`.

File: api/envoy/api/v2/listener/listener.proto (hunk 12 above)
Comment (mattklein123) [reply api/envoy/api/v2/listener/listener.proto]: Yeah I don't have a strong opinion on the order of IP vs. port. Perhaps ship and iterate? I'm fine either way.

File: api/envoy/api/v2/listener/listener.proto [hunk 23]
```diff
@@ -108,14 +110,12 @@ message FilterChainMatch {
...
   // ignored.
-  // [#not-implemented-hide:]
   repeated core.CidrRange source_prefix_ranges = 6;
 
   // The criteria is satisfied if the source port of the downstream connection
   // is contained in at least one of the specified ports. If the parameter is
   // not specified, the source port is ignored.
-  // [#not-implemented-hide:]
-  repeated google.protobuf.UInt32Value source_ports = 7;
+  repeated uint32 source_ports = 7 [(validate.rules).repeated .items.uint32 = {gte: 1, lte: 65535}];
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/api/v2/listener/listener.proto]: Technically no, but this field was never used or implemented, so I think it's fine to change?

File: docs/root/configuration/best_practices/best_practices.rst [hunk 24]
```diff
@@ -0,0 +1,7 @@
+Configuration best practices
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply docs/root/configuration/best_practices/best_practices.rst]: I think you are looking at a merge commit?

File: docs/root/configuration/configuration.rst [hunk 25]
```diff
@@ -5,7 +5,6 @@ Configuration reference
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply docs/root/configuration/configuration.rst]: I think you are looking at a merge commit?

File: source/server/listener_manager_impl.h [hunk 26]
```diff
@@ -333,56 +333,79 @@ class ListenerImpl : public Network::ListenerConfig,
   SystemTime last_updated_;
 
 private:
-  typedef std::array<Network::FilterChainSharedPtr, 3> SourceTypesArray;
-  typedef std::unordered_map<std::string, SourceTypesArray> ApplicationProtocolsMap;
-  typedef std::unordered_map<std::string, ApplicationProtocolsMap> TransportProtocolsMap;
+  using SourcePortsMap = absl::flat_hash_map<uint16_t, Network::FilterChainSharedPtr>;
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/server/listener_manager_impl.h]: You mean the switch to flat_hash_map? Or using? re: using that's what clang-tidy asks for in new code so I just switched it since I was switching to flash_hash_map also.

**Comment by mattklein123** [comment]
@PiotrSikora thanks for the review. Updated.

File: api/envoy/api/v2/listener/listener.proto (hunk 23 above)
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply api/envoy/api/v2/listener/listener.proto]: It looks that [go-control-plane](https://github.com/envoyproxy/go-control-plane) pulls those unimplemented fields anyway, and it already has [`source_ports`](https://github.com/envoyproxy/go-control-plane/blob/e90fc65c48c6e270d72ab798e0d7c441fcc0a807/envoy/api/v2/listener/listener.pb.go#L291) with type `[]*types.UInt32Value`, whereas this changes its type to `[]uint32`, which would break things (I think?).

cc @kyessenov

File: docs/root/configuration/best_practices/best_practices.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply docs/root/configuration/best_practices/best_practices.rst]: Nope, that was a comment that I made a week ago, when I first looked at this PR, and apparently it got stashed somewhere, even though it wasn't visible during today's round. Weird.

File: source/server/listener_manager_impl.h (hunk 26 above)
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply source/server/listener_manager_impl.h]: I was asking about `using`, thanks for the explanation.

File: api/envoy/api/v2/listener/listener.proto (hunk 23 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/api/v2/listener/listener.proto]: It would break things only if someone was actually using that field, though I can't imagine they would be. I can change it back if you want, just let me know.

[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/api/v2/listener/listener.proto]: Prefer not to do any more language breaking API changes in v2. Maybe tag with a `TODO` for v3, we can do these cleanups there.

File: source/server/listener_manager_impl.cc (hunk 20 above)
Comment (htuch) [reply source/server/listener_manager_impl.cc]: Just that this merges IP/port but that isn't done in dest IP/port code, it's fine as is.

File: api/envoy/api/v2/listener/listener.proto (hunk 12 above)
Comment (htuch) [reply api/envoy/api/v2/listener/listener.proto]: Thanks for the explanation @PiotrSikora. I wonder if we should capture this somewhere, it's useful design rationale.

File: api/envoy/api/v2/listener/listener.proto (hunk 23 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/api/v2/listener/listener.proto]: Will revert back.

[the remark this answers is not in the document]
Comment (kyessenov) [reply api/envoy/api/v2/listener/listener.proto]: I find it confusing to have experimental fields in stable protos, and I wouldn't discount that someone might be already using those fields (hopefully not in JSONs/YAMLs). It is more appropriate to fix those experimental fields in v3, but I understand the cost of branching the entire parent API version.

[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/api/v2/listener/listener.proto]: Actually, since this was `not-implemented-hide`, it's fair to break, even under the new stable API policy.

[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/api/v2/listener/listener.proto]: @kyessenov see https://github.com/envoyproxy/envoy/issues/6271#issuecomment-487716345, we are heading towards major versions.

[the remark this answers is not in the document]
Comment (mattklein123) [reply api/envoy/api/v2/listener/listener.proto]: OK, given that I can't get validations to work properly w/ the WKT, I'm going to go ahead and merge as-is.
